lib/SILOptimizer/Transforms/SimplifyCFG.cpp

```diff
@@ -1878,14 +1878,38 @@ static bool onlyForwardsNone(SILBasicBlock *noneBB, SILBasicBlock *someBB,
 ///    \            ... (more bbs?)
 ///     \           /
 ///       ulimateBB
-static bool hasSameUlitmateSuccessor(SILBasicBlock *noneBB, SILBasicBlock *someBB) {
+static bool hasSameUltimateSuccessor(SILBasicBlock *noneBB, SILBasicBlock *someBB) {
+  // Make sure that both our some, none blocks both have single successors that
+  // are not themselves (which can happen due to single block loops).
   auto *someSuccessorBB = someBB->getSingleSuccessorBlock();
-  if (!someSuccessorBB)
+  if (!someSuccessorBB || someSuccessorBB == someBB)
     return false;
   auto *noneSuccessorBB = noneBB->getSingleSuccessorBlock();
-  while (noneSuccessorBB != nullptr && noneSuccessorBB != someSuccessorBB)
-    noneSuccessorBB = noneSuccessorBB->getSingleSuccessorBlock();
-  return noneSuccessorBB == someSuccessorBB;
+  if (!noneSuccessorBB || noneSuccessorBB == noneBB)
+    return false;
+
+  // If we immediately find a diamond, return true. We are done.
+  if (noneSuccessorBB == someSuccessorBB)
+    return true;
+
+  // Otherwise, lets keep looking down the none case.
+  auto *next = noneSuccessorBB;
+  while (next != someSuccessorBB) {
+    noneSuccessorBB = next;
+    next = noneSuccessorBB->getSingleSuccessorBlock();
+
+    // If we find another single successor and it is not our own block (due to a
+    // self-loop), continue.
+    if (next && next != noneSuccessorBB)
+      continue;
+
+    // Otherwise, we either have multiple successors or a self-loop. We do not
```

Comment:
This fix only handles the special case of a single-block self loop. If SimplifyCFG ends up in a state with a multi-block self loop (e.g. if the loop blocks are not merged yet), it still hangs here. To reproduce, disable block merging

```
@@ -1304,7 +1304,7 @@ bool SimplifyCFG::simplifyBranchBlock(BranchInst *BI) {
 
   // If this block branches to a block with a single predecessor, then
   // merge the DestBB into this BB.
-  if (BB != DestBB && DestBB->getSinglePredecessorBlock()) {
+  if (BB != DestBB && DestBB->getSinglePredecessorBlock() && false) {
     LLVM_DEBUG(llvm::dbgs() << "merge bb" << BB->getDebugID() << " with bb"
                             << DestBB->getDebugID() << '\n');
```

and run SimplifyCFG on

```
sil @infinite_loop_1 : $@convention(thin) () -> @owned Optional<NSObject> {
bb0:
  %0 = function_ref @infinite_loop_get_optional_nsobject : $@convention(thin) () -> @autoreleased Optional<NSObject>
  %1 = apply %0() : $@convention(thin) () -> @autoreleased Optional<NSObject>
  br bb1(%1 : $Optional<NSObject>)

bb1(%3 : $Optional<NSObject>):
  switch_enum %3 : $Optional<NSObject>, case #Optional.some!enumelt.1: bb3, case #Optional.none!enumelt: bb2

bb2:
  br bb5(%3 : $Optional<NSObject>)

bb3(%7 : $NSObject):
  br bb4(%3 : $Optional<NSObject>)


bb4(%9 : $Optional<NSObject>):
  %11 = enum $Optional<NSObject>, #Optional.none!enumelt
  return %11 : $Optional<NSObject>

bb5(%14 : $Optional<NSObject>):
  br bb6

bb6:
  br bb5(%14 : $Optional<NSObject>)
}
```

A more robust fix - and simpler - fix would be to return false after e.g. 3 iterations of the while-loop (in the original version).

Reply:
A few things:

1. This was purposefully only fixing the single self-loop case. This is the bug that was reported.
2. Can you come up with a test case that doesnt hit the block merging path that exposes the large test case?
3. Are you suggesting just bailing after a number of iterations?

Reply:
If we think that we can get into that larger state, then rather than using a hard iteration limit, we should just use a small ptr set. That being said, I need to read the code again.

Reply:
Also, rather than making this change here initially. I would rather do any extra changes on master first and then cherry-pick separately.

Reply:
*separately == separate commits in same PR.

Reply:
You know what. I don't think it really matters whether or not we can reproduce this from SILGen. I think I can construct a SIL test case that will expose the multiple bb case without any options/etc (noting that SILGen will not be able to reproduce such a test case easily).